cpp/src/arrow/acero/hash_join.cc

```diff
@@ -306,12 +307,33 @@ class HashJoinBasicImpl : public HashJoinImpl {
 
     size_t num_probed_rows = match.size() + no_match.size();
     if (mask.is_scalar()) {
-      const auto& mask_scalar = mask.scalar_as<BooleanScalar>();
-      if (mask_scalar.is_valid && mask_scalar.value) {
-        // All rows passed, nothing left to do
-        return Status::OK();
+#if ARROW_LITTLE_ENDIAN
+       const auto& mask_scalar = mask.scalar_as<BooleanScalar>();
+       if (mask_scalar.is_valid && mask_scalar.value) {
+         // All rows passed, nothing left to do
+         return Status::OK();
+#else
+      // Check if the scalar is a BooleanScalar before casting
+      if (mask.scalar()->type->id() == Type::BOOL) {
```

Comment:
Why do we need this check? This check didn't exist in the existing code.

Reply:
This is a very detailed explanation. Thanks @Vishwanatha-HD .

As the author of the questioning test, I can say this is the issue of the test rather than of the hash join code. That is, using an filter expression evaluating to null-type `null` is invalid - it should be a boolean `null`. The hash join code itself arbitrarily assuming the expression being `boolean` is OK, though a `DCHECK_EQ(mask.type()->id(), Type::BOOL)` would be more preferable.

I think we should in turn fix the test by simply replacing the `literal(NullScalar())` with a boolean `null` - `literal(MakeNullScalar(boolean()))`

Reply:
Should be fixed by #48272 . @Vishwanatha-HD Could you first revert the change to `hash_join.cc` then rebase main to see if the hash join test is passing on big-endian? Thanks.